### PR TITLE
FTR-15: Reviews

### PR DESCRIPTION
**Pull request by mmarkoovin21**
# What is done

### Backend:
- **FirestoreReviews** class for fetching and adding reviews to the database

### Frontend:
-  **ReviewCard** component for displaying single review
-  **ReviewList** list that takes list of `**ReviewCard**`components and displays them inside `**CompanyDetailsContent**`
-  **AddReviewScreen** for adding reviews to database

# Notes

- App is storing multiple reviews from one user for one company. That should be resolved in the future.
- Should correct star icons on `AddReviewScreen` to not be filled at the beginning
- Should correct color on the back of `ReviewList`
- The user must put the rating,, that is, mark the stars

### CODE REVIEW

File: app/src/main/java/hr/foi/air/servicesync/data/Review.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package hr.foi.air.servicesync.data
+
+data class Review(
+    val companyId: String = "",
+    val description: String = "",
+    val rating: Int = 0,
+    val userId: String = ""
+)
```
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/data/Review.kt:8]: This is what I like to see.

File: app/src/main/java/hr/foi/air/servicesync/backend/FirestoreReviews.kt
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/backend/FirestoreReviews.kt:1]: Nice job for loggin info!

File: app/src/main/java/hr/foi/air/servicesync/backend/FirestoreReviews.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            }
+            .addOnFailureListener { exception ->
+                Log.e("FirestoreReviews", "Error fetching all reviews", exception)
+                onResult(emptyList())
+            }
+    }
+
+    fun fetchReviewsForCompany(companyId: String, onResult: (List<Review>) -> Unit) {
+        db.collection("reviews")
+            .whereEqualTo("companyId", companyId.trim())
```
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/backend/FirestoreReviews.kt:13]: Trimming, good thinking!

File: app/src/main/java/hr/foi/air/servicesync/navigation/AppNavHost.kt [hunk 3]
```diff
@@ -82,4 +82,30 @@ fun NavGraphBuilder.AppNavHost(navController: NavHostController) {
...
+                        context,
+                        "Review added successfully.",
+                        Toast.LENGTH_SHORT
+                    ).show()
+                } else {
+                    Toast.makeText(
+                        context,
+                        "Failed to add review. Please try again.",
+                        Toast.LENGTH_SHORT
+                    ).show()
```
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/navigation/AppNavHost.kt:106]: I don't see why appnavhost should know what to do when a review is submitted or failed.

File: app/src/main/java/hr/foi/air/servicesync/ui/components/ReviewCard.kt
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/ui/components/ReviewCard.kt:1]: Perfect!

File: app/src/main/java/hr/foi/air/servicesync/ui/contents/CompanyDetailsContent.kt [hunk 4]
```diff
@@ -45,12 +52,14 @@ fun CompanyDetailsContent(
     companyName: String
 ) {
     val firestoreCompanyDetails = FirestoreCompanyDetails()
+    val firestoreReviews = FirestoreReviews()
```
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/ui/contents/CompanyDetailsContent.kt]: I am not a fan of this. This breaks three-layer architecture.

File: app/src/main/java/hr/foi/air/servicesync/ui/contents/CompanyDetailsContent.kt [hunk 5]
```diff
@@ -75,6 +84,10 @@ fun CompanyDetailsContent(
         firestoreCompanyDetails.loadCompanyImageUrlByName(companyName) { imageUrl ->
             companyImageUrl.value = imageUrl
         }
+        firestoreReviews.fetchReviewsForCompany(companyName) { fetchedReviews ->
+            Log.d("COMPANYNAME", companyName)
+            reviews.value = fetchedReviews
+        }
```
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/ui/contents/CompanyDetailsContent.kt]: Again, breaks three layer architecutre. Consider createing a class e.g. ReviewHandler that uses firestore and returns to the UI all the reviews

File: app/src/main/java/hr/foi/air/servicesync/ui/contents/CompanyDetailsContent.kt [hunk 6]
```diff
@@ -161,25 +174,33 @@ fun CompanyDetailsContent(
...
+                        verticalAlignment = Alignment.CenterVertically
+                    ) {
+                        Text(
+                            text = stringResource(id = R.string.reviews),
+                            color = isDark(onSurfaceDark, onSurfaceLight),
+                            style = MaterialTheme.typography.headlineMedium
+                        )
+                        Button(
+                            onClick = {
+                                navController.navigate("addReview/$companyName/${UserSession.username}")
```
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/ui/contents/CompanyDetailsContent.kt:189]: Nicely done for navigating to another screen.

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+@Composable
+fun AddReviewScreen(
+    modifier: Modifier = Modifier,
+    companyId: String,
+    userId: String,
+    navController: NavController,
+    onReviewSubmit: (Boolean) -> Unit
+) {
+    val firestoreReviews = FirestoreReviews()
```
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt:41]: Not a fan.

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+    Column(
+        modifier = modifier
+            .fillMaxSize()
+            .padding(16.dp),
+        verticalArrangement = Arrangement.spacedBy(16.dp),
+        horizontalAlignment = Alignment.CenterHorizontally
+    ) {
+        Text(
+            text = "Dodavanje recenzije",
```
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt]: Static text not in strings.xml? Add it

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+                    companyId = companyId,
+                    description = reviewText.value,
+                    rating = rating.value,
+                    userId = userId
+                )
+
+                firestoreReviews.addReview(review) { success ->
+                    onReviewSubmit(success)
+                }
+                navController.navigate("company/$companyId")
```
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt]: Nice. Just as I was expecting, after the view, it redirects to the company details. However, this maybe breaks the stack of previous screens. Maybe you could just pop the stack.

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+            },
+            modifier = Modifier.fillMaxWidth()
+        ) {
+            Text(text = stringResource(R.string.po_alji_recenziju))
```
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt]: String resources should be in english

**Review by mpuskadij — APPROVED** [review]
Main considerations:

- some static text is not in strings.xml
- some composable functions break three layer architecture

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt (hunk 7 above)
Comment (mmarkoovin21) [reply app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt:41]: Will refactor in other sprint.

File: app/src/main/java/hr/foi/air/servicesync/navigation/AppNavHost.kt (hunk 3 above)
Comment (mmarkoovin21) [reply app/src/main/java/hr/foi/air/servicesync/navigation/AppNavHost.kt:106]: I added it just for info when review is submitted.

File: app/src/main/java/hr/foi/air/servicesync/ui/components/ReviewCard.kt [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+import hr.foi.air.servicesync.data.Review
+
+@Composable
+fun ReviewCard(
+    review: Review
+) {
+    Card(
+        shape = RoundedCornerShape(8.dp),
+        colors = CardDefaults.cardColors(
+            containerColor = MaterialTheme.colorScheme.surfaceVariant,
```
Comment (dblazevic21) [inline app/src/main/java/hr/foi/air/servicesync/ui/components/ReviewCard.kt:33]: This should use colors from Colors.kt. Change to the color palette of Figma if not the same hex code

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt (hunk 9 above)
Comment (dblazevic21) [reply app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt]: Will probably work better with popBackStack()

Comment (mmarkoovin21) [reply app/src/main/java/hr/foi/air/servicesync/ui/screens/AddReviewScreen.kt]: Corrected!

File: app/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -14,7 +14,9 @@
     <string name="reviews">Recenzije</string>
     <string name="description">Opis usluge</string>
     <string name="logo_text">Sve na jednom mjestu</string>
+    <string name="star">Star</string>
     <string name="confirm">Potvrdi</string>
     <string name="cancel">Odustani</string>
     <string name="select_date">Odaberite datum...</string>
+    <string name="po_alji_recenziju">Pošalji recenziju</string>
```
Comment (dblazevic21) [inline app/src/main/res/values/strings.xml]: Make this string name "submit_review" and not "po_alji_recenziju"

**Review by dblazevic21 — CHANGES_REQUESTED** [review]
Maybe also remove these Logs

File: app/src/main/java/hr/foi/air/servicesync/ui/components/ReviewCard.kt (hunk 11 above)
Comment (mmarkoovin21) [reply app/src/main/java/hr/foi/air/servicesync/ui/components/ReviewCard.kt:33]: Sure, but then the rest of the screen is in other color. We need to coordinate those colors better.

**Pull request merged**